src/NumericalAlgorithms/Spectral/Spectral.hpp

```diff
@@ -0,0 +1,143 @@
+ * a template-specialized constructor.
+ */
+
+enum class Basis { Legendre };
+
+/*!
+ * \ingroup SpectralGroup
+ * \brief The choice of quadrature method to compute integration weights.
+ *
+ * \details Integrals using \f$\f$N collocation points with Gauss quadrature are
+ * exact to polynomial order \f$p=2N-1\f$. Gauss-Lobatto quadrature is exact
+ * only to polynomial order \f$p=2N-3\f$, but includes collocation points at the
+ * domain boundary.
+ *
+ * \note For each member of `Spectral::Basis` and `Spectral::Quadrature`,
+ * the class `Spectral::detail::CollocationPointsAndWeights` should implement
+ * a template-specialized constructor.
+ */
+enum class Quadrature { Gauss, GaussLobatto };
+
+/*!
+ * \brief Collocation points
+ * \param number_of_pts number of collocation points
+ */
+template <Basis BasisType, Quadrature QuadratureType>
+const DataVector& collocation_points(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Quadrature weights
+ * \param number_of_pts number of collocation points
+ */
+template <Basis BasisType, Quadrature QuadratureType>
+const DataVector& quadrature_weights(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to compute the first derivative of a function.
+ * \param number_of_pts number of collocation points
+ */
+template <Basis BasisType, Quadrature QuadratureType>
+const Matrix& differentiation_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to compute spectral coefficients of a function.
+ * \param number_of_pts number of collocation points
+ */
+template <Basis BasisType, Quadrature QuadratureType>
+const Matrix& grid_points_to_spectral_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to compute function at the collocation points from its
+ * spectral coefficients.
+ * \param number_of_pts number of collocation points
+ */
+template <Basis BasisType, Quadrature QuadratureType>
+const Matrix& spectral_to_grid_points_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to linearize a function.
+ *
+ * \details Filters out all except the lowest two modes.
+ * \param number_of_pts number of collocation points
+ */
+template <Basis BasisType, Quadrature QuadratureType>
+const Matrix& linear_filter_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix to interpolate to a set of target points
+ *
+ * \warning It is expected but not checked that the target points are inside
+ * the interval covered by the `BasisType` in logical coordinates.
+ * \param number_of_pts The number of collocation points
+ * \param target_points The points to interpolate to
+ */
+template <Basis BasisType, Quadrature QuadratureType, typename T>
+Matrix interpolation_matrix(size_t number_of_pts,
+                            const T& target_points) noexcept;
+
+/*!
+ * \brief Maximum number of allowed collocation points
+ */
+constexpr size_t maximum_number_of_pts = 12;
+
+namespace detail {
```

Comment:
As this is detail, please hide it in a cpp file;  our suggestion is to have collocation_points and quadrature_weights be defined in the specific basis files

Reply:
Right now `Spectral.cpp` computes basis-agnostic quantities, takes care of the caching and provides accessors to everything, whereas `Legendre.cpp` implements basis-specific algorithms. Since `Spectral.cpp` needs to cache the results of these functions, I think I can't just let `Legendre.cpp` implement `Spectral::collocation_points` and `Spectral::quadrature_weights` directly (or else caching code would be duplicated). So I don't quite see how to get rid of the `detail` namespace. Perhaps we could talk about this in a hangout next week?

Reply:
Thanks for doing the split while I was asleep :) 👍